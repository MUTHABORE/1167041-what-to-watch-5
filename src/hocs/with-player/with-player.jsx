import React from 'react';
import {PureComponent} from 'react';
import {connect} from 'react-redux';
import PropTypes from 'prop-types';
import {propsForFilms, propsForRouterProps} from '../../util/props-validation.js';

export const withPlayer = (Component) => {
  class WithPlayer extends PureComponent {
    constructor(props) {
      super(props);

      this.state = {
        isPlaying: true,
        durationProgress: 0,
        durationLeft: 0,
      };

      this.videoRef = React.createRef();

      this.moviesList = props.moviesList;
      this.routerProps = props.routerProps;
      this.movie = this.moviesList.find((elem) => elem.id === +this.routerProps.match.params.id);

      this.playerStatusChangeHandler = this.playerStatusChangeHandler.bind(this);
      this.closeButtonClickHandler = this.closeButtonClickHandler.bind(this);
      this.durationUpdateHandler = this.durationUpdateHandler.bind(this);
      this.fullscreenClickHandler = this.fullscreenClickHandler.bind(this);
    }

    componentDidMount() {
      const video = this.videoRef.current;
      video.play();
    }

    componentDidUpdate() {
      const video = this.videoRef.current;
      if (this.state.isPlaying === true) {
        video.play();
      } else {
        video.pause();
      }
    }

    playerStatusChangeHandler() {
      this.setState({isPlaying: !this.state.isPlaying});
    }

    closeButtonClickHandler() {
      this.routerProps.history.goBack();
    }

    durationUpdateHandler() {
      const video = this.videoRef.current;

      this.setState({
        durationProgress: video.currentTime * 100 / video.duration,
        durationLeft: video.duration - video.currentTime,
      });
    }

    fullscreenClickHandler() {
      const video = this.videoRef.current;
      video.requestFullscreen();
    }

    render() {
      return (
        <Component
          {...this.props}
          movie={this.movie}
          playerStatusChangeHandler={this.playerStatusChangeHandler}
          closeButtonClickHandler={this.closeButtonClickHandler}
          durationProgress={this.state.durationProgress}
          durationLeft={this.state.durationLeft}
          fullscreenClickHandler={this.fullscreenClickHandler}
          isPlaying={this.state.isPlaying}
        >
          <video
            ref={this.videoRef}
            className="player__video"
            src={this.movie.video_link}
            poster={this.movie.preview_image}
            onTimeUpdate={this.durationUpdateHandler}
            loop
            muted
          >
          </video>
        </Component>
      );
    }
  }

  WithPlayer.propTypes = {
    moviesList: PropTypes.arrayOf(propsForFilms).isRequired,
    routerProps: propsForRouterProps,
  };

  const mapStateToProps = (state) => ({
    moviesList: state.DATA.moviesList,
  });

  return connect(mapStateToProps)(WithPlayer);
};
