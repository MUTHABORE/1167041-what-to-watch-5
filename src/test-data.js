import React from 'react';
import PropTypes from 'prop-types';

export const noop = () => {};

export const routerProps = {
  match: {
    params: {
      id: `1`,
    }
  }
};

export const movies = [
  {
    name: `Bronson`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    rating: 3.6,
    votes: 109661,
    director: `Nicolas Winding Refn`,
    starring: [
      `Tom Hardy`,
      `Kelly Adams`,
      `Luing Andrews`
    ],
    runtime: 92,
    genre: `Action`,
    released: 2008,
    id: 1,
    isFavorite: false,
    video: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  },
  {
    name: `Bronson`,
    poster: `https://assets.htmlacademy.ru/intensives/javascript-3/film/poster/bronson.jpg`,
    preview: `https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg`,
    background: `https://assets.htmlacademy.ru/intensives/javascript-3/film/background/bronson.jpg`,
    backgroundColor: `#73B39A`,
    description: `A young man who was sentenced to seven years in prison for robbing a post office ends up spending three decades in solitary confinement. During this time, his own personality is supplanted by his alter-ego, Charles Bronson.`,
    rating: 3.6,
    votes: 109661,
    director: `Nicolas Winding Refn`,
    starring: [
      `Tom Hardy`,
      `Kelly Adams`,
      `Luing Andrews`
    ],
    runtime: 92,
    genre: `Action`,
    released: 2008,
    id: 2,
    isFavorite: true,
    video: `http://media.xiph.org/mango/tears_of_steel_1080p.webm`,
    previewVideo: `https://upload.wikimedia.org/wikipedia/commons/transcoded/b/b3/Big_Buck_Bunny_Trailer_400p.ogv/Big_Buck_Bunny_Trailer_400p.ogv.360p.webm`
  }
];

export const reviews = [
  {
    id: 1,
    user: {
      id: 11,
      name: `Jack`
    },
    rating: 8.6,
    comment: `A movie that will take you to another world full of emotions.`,
    date: `2020-11-03T13:38:44.769Z`
  },
  {
    id: 2,
    user: {
      id: 12,
      name: `Jack`
    },
    rating: 8.6,
    comment: `A movie that will take you to another world full of emotions.`,
    date: `2020-11-03T13:38:44.769Z`
  }
];

export const movie = movies[0];

export const userMock = {
  id: 1,
  name: `GERASIM`,
  email: `moo@moo.com`,
  avatarUrl: `https://assets.htmlacademy.ru/intensives/javascript-3/avatar/5.jpg`,
};

export const state = {
  DATA: {
    moviesList: [movies[0], movies[1]],
    genresList: [`All genres`, `Action`],
    reviews: [],
    promo: movies[0],
  },
  FUNCTIONAL: {
    activeGenre: `All genres`,
    amountMoviesToRender: 8,
  },
  USER: {
    authorizationStatus: `NO_AUTH`,
    user: userMock,
  },
};

export const MockComponent = (props) => {
  const {children} = props;
  return (
    <div>
      {children}
    </div>
  );
};

MockComponent.propTypes = {
  children: PropTypes.oneOfType([
    PropTypes.arrayOf(PropTypes.node),
    PropTypes.node
  ]).isRequired,
};

export const SimpleMockComponent = () => <div />;
