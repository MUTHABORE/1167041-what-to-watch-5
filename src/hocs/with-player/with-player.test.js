import React from 'react';
import renderer from 'react-test-renderer';
import {Provider} from 'react-redux';
import configurateMockState from 'redux-mock-store';
import {MockComponent, state, routerProps, noop} from '../../test-data';
import {withPlayer} from './with-player.jsx';

const MockComponentWrapped = withPlayer(MockComponent);

const mockStore = configurateMockState()(state);

test(`Should withPlayer render correctly`, () => {
  const tree = renderer
    .create((
      <Provider store={mockStore}>
        <MockComponentWrapped
          moviesList={state.DATA.moviesList}
          routerProps={routerProps}
        >
          <React.Fragment />
        </MockComponentWrapped>
      </Provider>
    ), {
      createNodeMock() {
        return {
          play: noop,
          pause: noop,
        };
      }
    })
    .toJSON();

  expect(tree).toMatchSnapshot();
});
