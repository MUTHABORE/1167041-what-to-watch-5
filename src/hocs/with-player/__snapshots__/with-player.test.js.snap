// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Should withPlayer render correctly 1`] = `
<div>
  <video
    className="player__video"
    loop={true}
    muted={true}
    onTimeUpdate={[Function]}
    poster="https://assets.htmlacademy.ru/intensives/javascript-3/film/preview/bronson.jpg"
    src="http://media.xiph.org/mango/tears_of_steel_1080p.webm"
  />
</div>
`;
